// @heronjs/common/decorators/storage/storage.decorator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageClass = exports.StorageProtocol = void 0;
const reflector_1 = require("../../reflector");
const declarations_1 = require("../../declarations");
var StorageProtocol;
(function (StorageProtocol) {
    StorageProtocol[StorageProtocol["S3"] = 0] = "S3";
    StorageProtocol[StorageProtocol["Azure"] = 1] = "Azure";
    StorageProtocol[StorageProtocol["GCS"] = 2] = "GCS";
    StorageProtocol[StorageProtocol["FTP"] = 3] = "FTP";
    StorageProtocol[StorageProtocol["LOCAL"] = 4] = "LOCAL";
})(StorageProtocol = exports.StorageProtocol || (exports.StorageProtocol = {}));
const StorageClass = (clazz) => {
    return target => {
        const storages = (0, reflector_1.__obtain)(declarations_1.MODULE_STORAGES, target);
        storages && storages.push(clazz);
        (0, reflector_1.__mixinIfAbsent)(target, { name: declarations_1.MODULE_STORAGES, refers: () => storages });
    };
};
exports.StorageClass = StorageClass;
