// @heronjs/common/utils/validators/class.validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClassValidator = void 0;
const tslib_1 = require("tslib");
const class_validator_1 = require("class-validator");
class ClassValidator {
    static validate(entity, context) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            let details;
            const errors = yield (0, class_validator_1.validate)(entity, {
                skipMissingProperties: true
            });
            if (errors && errors.length > 0) {
                details = {
                    context: context || entity.constructor.name,
                    errors: []
                };
                const takeFirst = false;
                for (const e of errors) {
                    const messages = e.constraints ? Object.values(e.constraints) : [];
                    let error;
                    details.errors.push({
                        property: e.property,
                        message: messages
                    });
                    if (messages && messages.length > 0 && !takeFirst) {
                        error = messages[0];
                        details.error = error;
                    }
                }
            }
            return details;
        });
    }
}
exports.ClassValidator = ClassValidator;
