// @heronjs/common/logger/logger.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMBINED_TRANSPORT_CONSOLE = exports.DEV_TRANSPORT = exports.PROD_TRANSPORT_CONSOLE = exports.PROD_TRANSPORT_FILE = exports.DEFAULT_FORMATTER = exports.DEFAULT_LEVEL = void 0;
const tslib_1 = require("tslib");
const winston_1 = tslib_1.__importStar(require("winston"));
exports.DEFAULT_LEVEL = {
    levels: {
        trace: 5,
        debug: 4,
        info: 3,
        warn: 2,
        error: 1,
        fatal: 0
    },
    colors: {
        trace: 'white',
        debug: 'green',
        info: 'green',
        warn: 'yellow',
        error: 'red',
        fatal: 'red'
    }
};
const DEFAULT_FORMATTER = (name) => winston_1.default.format.combine(winston_1.format.errors({ stack: true }), winston_1.format.label({ label: name }), winston_1.format.colorize(), winston_1.format.timestamp({ format: 'MM-DD-YYYY HH:mm:ss' }), winston_1.format.splat(), winston_1.format.printf((info) => {
    const { label, timestamp, level, message } = info, meta = tslib_1.__rest(info, ["label", "timestamp", "level", "message"]);
    return `${timestamp} ${label} [${level}]: ${message} ${Object.keys(meta).length ? JSON.stringify(meta, null, 2) : ''}`;
}));
exports.DEFAULT_FORMATTER = DEFAULT_FORMATTER;
exports.PROD_TRANSPORT_FILE = new winston_1.default.transports.File({
    filename: 'logs/error.log',
    level: 'error'
});
const PROD_TRANSPORT_CONSOLE = (name) => new winston_1.default.transports.Console({
    format: (0, exports.DEFAULT_FORMATTER)(name)
});
exports.PROD_TRANSPORT_CONSOLE = PROD_TRANSPORT_CONSOLE;
exports.DEV_TRANSPORT = new winston_1.default.transports.Console({
    format: (0, exports.DEFAULT_FORMATTER)('dev')
});
const COMBINED_TRANSPORT_CONSOLE = (name) => [
    new winston_1.default.transports.Console({ format: (0, exports.DEFAULT_FORMATTER)(name) }),
    new winston_1.default.transports.File({ filename: './logs/server.log' })
];
exports.COMBINED_TRANSPORT_CONSOLE = COMBINED_TRANSPORT_CONSOLE;
