// @heronjs/common/consts/database.consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DatabaseClient = exports.DatabaseDriver = exports.ConnectionMode = exports.MODULE_DATABASE = void 0;
/**
 * runtime module database token
 */
exports.MODULE_DATABASE = 'ModuleDatabase';
/**
 * Database connection mode
 */
var ConnectionMode;
(function (ConnectionMode) {
    ConnectionMode[ConnectionMode["STANDALONE"] = 0] = "STANDALONE";
    ConnectionMode[ConnectionMode["CLUSTER"] = 1] = "CLUSTER";
})(ConnectionMode = exports.ConnectionMode || (exports.ConnectionMode = {}));
/**
 * database driver
 */
var DatabaseDriver;
(function (DatabaseDriver) {
    DatabaseDriver["MYSQL"] = "mysql";
    DatabaseDriver["POSTGRES"] = "postgres";
    DatabaseDriver["SQLITE"] = "sqlite3";
    DatabaseDriver["ORACLE"] = "oracledb";
    DatabaseDriver["MARIADB"] = "mariadb";
    DatabaseDriver["MONGODB"] = "mongodb";
    DatabaseDriver["CASSANDRA"] = "cassandra";
})(DatabaseDriver = exports.DatabaseDriver || (exports.DatabaseDriver = {}));
/**
 * database client
 */
var DatabaseClient;
(function (DatabaseClient) {
    DatabaseClient[DatabaseClient["KNEX"] = 0] = "KNEX";
})(DatabaseClient = exports.DatabaseClient || (exports.DatabaseClient = {}));
