// @heronjs/common/consts/errors.consts.js
"use strict";
/**
 * error types
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SQLErrors = exports.ConstrainErrors = exports.RTErrors = exports.EVNErrors = exports.SYSErrors = exports.Errors = void 0;
var Errors;
(function (Errors) {
    Errors["SQL_ERR"] = "SQL-ERR";
    Errors["API_ERR"] = "API-ERR";
    Errors["SYS_ERR"] = "SYS-ERR";
    Errors["RUNTIME_ERR"] = "RUNTIME-ERR";
    Errors["VALIDATION_ERR"] = "VALIDATION-ERR";
})(Errors = exports.Errors || (exports.Errors = {}));
/**
 * system error
 * begin from 1000
 */
var SYSErrors;
(function (SYSErrors) {
    SYSErrors[SYSErrors["SERVER_ERR"] = 1000] = "SERVER_ERR";
    SYSErrors[SYSErrors["MODULE_ERR"] = 1001] = "MODULE_ERR";
    SYSErrors[SYSErrors["FUNCTION_ERR"] = 1002] = "FUNCTION_ERR";
    SYSErrors[SYSErrors["HTTP_SERVER_ERR"] = 1003] = "HTTP_SERVER_ERR";
    SYSErrors[SYSErrors["NOT_SUPPORTED_ERROR"] = 1004] = "NOT_SUPPORTED_ERROR";
    SYSErrors[SYSErrors["GRAPHQL_SERVER_ERR"] = 1005] = "GRAPHQL_SERVER_ERR";
})(SYSErrors = exports.SYSErrors || (exports.SYSErrors = {}));
/**
 * environment error
 * begin from 2000
 */
var EVNErrors;
(function (EVNErrors) {
    EVNErrors[EVNErrors["MESSAGE_HANDLE_ERR"] = 2000] = "MESSAGE_HANDLE_ERR";
    EVNErrors[EVNErrors["MESSAGE_REJECT_ERR"] = 2001] = "MESSAGE_REJECT_ERR";
})(EVNErrors = exports.EVNErrors || (exports.EVNErrors = {}));
/**
 * runtime error
 * begin from 3000
 */
var RTErrors;
(function (RTErrors) {
    RTErrors[RTErrors["NOT_IMPLEMENTED_ERR"] = 3000] = "NOT_IMPLEMENTED_ERR";
    RTErrors[RTErrors["NOT_HANDLED_ERR"] = 3001] = "NOT_HANDLED_ERR";
    RTErrors[RTErrors["DEFINE_META_ERR"] = 3002] = "DEFINE_META_ERR";
    RTErrors[RTErrors["INJECTION_ERR"] = 3003] = "INJECTION_ERR";
})(RTErrors = exports.RTErrors || (exports.RTErrors = {}));
/**
 * Validator Error
 */
var ConstrainErrors;
(function (ConstrainErrors) {
    ConstrainErrors[ConstrainErrors["INVALID_CONSTRAIN_ERR"] = 4000] = "INVALID_CONSTRAIN_ERR";
    ConstrainErrors[ConstrainErrors["INVALID_DEFINE_ERR"] = 4001] = "INVALID_DEFINE_ERR";
})(ConstrainErrors = exports.ConstrainErrors || (exports.ConstrainErrors = {}));
/**
 * sql error
 * begin from 5000
 */
var SQLErrors;
(function (SQLErrors) {
    SQLErrors[SQLErrors["DUPLICATE_ENTRY_ERR"] = 5000] = "DUPLICATE_ENTRY_ERR";
    SQLErrors[SQLErrors["ENTITY_EXISTS_ERR"] = 5001] = "ENTITY_EXISTS_ERR";
    SQLErrors[SQLErrors["ENTITY_NOT_FOUND_ERR"] = 5002] = "ENTITY_NOT_FOUND_ERR";
    SQLErrors[SQLErrors["CONSTRAINTS_ERR"] = 5003] = "CONSTRAINTS_ERR";
    SQLErrors[SQLErrors["INVALID_SQL_ERR"] = 5004] = "INVALID_SQL_ERR";
    SQLErrors[SQLErrors["CONNECTION_ERR"] = 5005] = "CONNECTION_ERR";
})(SQLErrors = exports.SQLErrors || (exports.SQLErrors = {}));
