// @heronjs/common/consts/values.consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SYSTEM_REGISTRY = exports.MODULE_PLUGINS = exports.DEFAULT_CACHE_HEADER = exports.CACHE_EVENT_HANDLER = exports.DEFAULT_UPLOAD_LIMIT_SIZE = exports.DEFAULT_LIMIT_HTTP_SIZE = exports.GATEKEEPER_RESOLVER = exports.GATEKEEPER = exports.MODULE_CONFIGURATION = exports.SYSTEM_VARIABLES = exports.SERVICE_CONTEXT_EVENTS = exports.SERVICE_CONTEXT = exports.EVENT_CONTEXT = exports.DEFAULT_LIMITED_MAP_VALUE = void 0;
exports.DEFAULT_LIMITED_MAP_VALUE = parseInt(process.env.MAX_OBJECT_POOL || '1000', 10);
exports.EVENT_CONTEXT = '$event-context';
exports.SERVICE_CONTEXT = '$service-context';
exports.SERVICE_CONTEXT_EVENTS = '$service-context-events';
exports.SYSTEM_VARIABLES = '$system-variables';
exports.MODULE_CONFIGURATION = 'module-configuration';
exports.GATEKEEPER = 'module-gatekeeper';
exports.GATEKEEPER_RESOLVER = 'module-gatekeeper-resolver';
exports.DEFAULT_LIMIT_HTTP_SIZE = process.env.LIMIT_HTTP_BODY_SIZE || '15mb';
exports.DEFAULT_UPLOAD_LIMIT_SIZE = process.env.LIMIT_UPLOAD_BODY_SIZE || '200mb';
exports.CACHE_EVENT_HANDLER = '$system-$ceh';
exports.DEFAULT_CACHE_HEADER = 'X-Cache';
exports.MODULE_PLUGINS = '$module_$plugins';
exports.SYSTEM_REGISTRY = '$system_registry$' + Date.now();
