// @heronjs/express/processors/abstract.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractProcessor = void 0;
const common_1 = require("@heronjs/common");
const module_consts_1 = require("../module.consts");
class AbstractProcessor {
    constructor() {
        this.withExtensions = (cacheStore, systemEvent, di) => {
            cacheStore && (this._cacheManager = cacheStore === null || cacheStore === void 0 ? void 0 : cacheStore.get());
            systemEvent && (this._systemEvent = systemEvent);
            di && (this._di = di);
        };
        this.$submitCacheEvent = (cv, request) => {
            (cv && this._systemEvent) && this._systemEvent.emit(common_1.SystemEventType.CACHING, {
                data: {
                    key: cv.key || (request === null || request === void 0 ? void 0 : request.url),
                    ttl: cv.ttl || module_consts_1.DEFAULT_CACHE_EXPIRY,
                    value: cv.value,
                    validate: cv.validate,
                    behavior: cv.behavior
                }
            });
        };
        this.$cacheIfAvailable = (result, request, cache) => {
            cache && this.$submitCacheEvent(Object.assign(Object.assign({}, cache), { key: cache.key ? cache.key : request.url, value: result }), request);
        };
    }
}
exports.AbstractProcessor = AbstractProcessor;
