// @heronjs/common/declarations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SELF_TRANSFORMABLE = exports.INSTANCE_TYPE = exports.INSTANCE_NAME = exports.ENTITY_COLUMN = exports.ENTITY_META = exports.SCOPE_OPTIONS_METADATA = exports.ROUTE_INTERCEPTORS = exports.ROUTE_GUARD = exports.FUNCTION_META = exports.FUNCTION_PARAMETERS = exports.OBJECT_FUNCTIONS = exports.OBJECT_RUNNER = exports.CTRL_GUARD = exports.CTRL_VIEW = exports.GRAPHQL_RESOLVERS = exports.GRAPHQL_TYPES = exports.MODULE_STORES = exports.CTRL_INTERCEPTORS = exports.CTRL_GRAPHQL = exports.CTRL_GRPC = exports.APP_CACHES = exports.CTRL_WEBSOCKET = exports.CTRL_ROUTES = exports.CTRL_PREFIX = exports.SELF_RESOLVER = exports.SELF_EVENT = exports.SELF_META = exports.MODULE_GRAPHQL = exports.MODULE_STORAGES = exports.MODULE_EXPORTS = exports.MODULE_IMPORTS = exports.MODULE_VIEWS = exports.DEFAULT_PROVIDER = exports.MODULE_PROVIDERS = exports.MODULE_GRPC = exports.MODULE_ID = exports.MODULE_AUTH_CONTEXT = exports.MODULE_WEBSOCKET = exports.MODULE_CONTROLLERS = exports.MODULE_GATEKEEPER = exports.MODULE_DAEMONS = exports.MODULE_SERVICES = exports.MODULE_DATABASES = exports.MODULE_DI = exports.MODULE_ROOT = exports.GLOBAL_MODULE_METADATA = exports.HERON_PLUGINS = exports.HERON_MODULES = exports.GLOBAL_IOC = exports.MODULE_METADATA = void 0;
exports.MODULE_ALIAS = exports.DESIGN_PARAM_TYPE = exports.DESIGN_RETURN_TYPE = exports.STORAGE_STORE = exports.CACHE_STORE = exports.SELF_CONFIG = void 0;
exports.MODULE_METADATA = {
    PROVIDERS: 'providers',
    CONTROLLERS: 'controllers',
    IMPORTS: 'imports',
    VIEWS: 'views',
    STORES: 'stores',
    SERVICES: 'services',
    WEBSOCKETS: 'websockets',
    GRAPHQLS: 'graphqls',
    EXPORTS: 'exports',
    ID: 'id'
};
/**
 * Heron metadata
 */
exports.GLOBAL_IOC = '__global:ioc__';
exports.HERON_MODULES = '__heron:modules__';
exports.HERON_PLUGINS = '__heron:plugins__';
exports.GLOBAL_MODULE_METADATA = '__module:global__';
exports.MODULE_ROOT = '__module:root__';
exports.MODULE_DI = '__module:di__';
exports.MODULE_DATABASES = '__module:databases__';
exports.MODULE_SERVICES = '__module:services';
exports.MODULE_DAEMONS = '__module:daemon__';
exports.MODULE_GATEKEEPER = '__module:gatekeeper__';
exports.MODULE_CONTROLLERS = '__module:controllers__';
exports.MODULE_WEBSOCKET = '__module:websocket__';
exports.MODULE_AUTH_CONTEXT = '__module:auth-context__';
exports.MODULE_ID = '__module:id__';
exports.MODULE_GRPC = '__module:grpc__';
exports.MODULE_PROVIDERS = '__module:providers__';
exports.DEFAULT_PROVIDER = '__provider:default__';
exports.MODULE_VIEWS = '__module:views__';
exports.MODULE_IMPORTS = '__module:imports__';
exports.MODULE_EXPORTS = '__module:exports__';
exports.MODULE_STORAGES = '__module:storages__';
exports.MODULE_GRAPHQL = '__module:graphql__';
exports.SELF_META = '__self:metadata__';
exports.SELF_EVENT = '__self:event.name__';
exports.SELF_RESOLVER = '__self:resolver__';
exports.CTRL_PREFIX = '__controller:prefix__';
exports.CTRL_ROUTES = '__controller:routes__';
exports.CTRL_WEBSOCKET = '__controller:websocket__';
exports.APP_CACHES = '__app:caches__';
exports.CTRL_GRPC = '__controller:grpc__';
exports.CTRL_GRAPHQL = '__controller:graphql__';
exports.CTRL_INTERCEPTORS = '__controller:interceptors__';
exports.MODULE_STORES = '__module:stores__';
exports.GRAPHQL_TYPES = '__graphql:types__';
exports.GRAPHQL_RESOLVERS = '__graphql:resolvers__';
exports.CTRL_VIEW = '__controller:view__';
exports.CTRL_GUARD = '__controller:guard__';
exports.OBJECT_RUNNER = '__controller:runners__';
exports.OBJECT_FUNCTIONS = '__object:functions__';
exports.FUNCTION_PARAMETERS = '__function:parameters__';
exports.FUNCTION_META = '__function:meta__';
exports.ROUTE_GUARD = '__route:guard__';
exports.ROUTE_INTERCEPTORS = '__route:interceptors__';
exports.SCOPE_OPTIONS_METADATA = '__scope:options__';
exports.ENTITY_META = '__entity:meta__';
exports.ENTITY_COLUMN = '__entity:column__';
exports.INSTANCE_NAME = '__instance:name__';
exports.INSTANCE_TYPE = '__instance:type__';
exports.SELF_TRANSFORMABLE = '__self:transformable__';
exports.SELF_CONFIG = '__self:config__';
exports.CACHE_STORE = '__store:cache__';
exports.STORAGE_STORE = '__store:storage__';
/**
 * Metadata reflection
 */
exports.DESIGN_RETURN_TYPE = 'design:returntype';
exports.DESIGN_PARAM_TYPE = 'design:paramtypes';
exports.MODULE_ALIAS = {
    controllers: exports.MODULE_CONTROLLERS,
    providers: exports.MODULE_PROVIDERS,
    imports: exports.HERON_MODULES,
    views: exports.MODULE_VIEWS,
    stores: exports.MODULE_STORES,
    services: exports.MODULE_SERVICES,
    websockets: exports.MODULE_WEBSOCKET,
    graphqls: exports.MODULE_GRAPHQL,
    exports: exports.MODULE_EXPORTS,
    id: exports.MODULE_ID
};
